applications/vanilla/models/class.discussionmodel.php

```diff
@@ -1144,10 +1144,17 @@ public function calculate(&$discussion) {
             $discussion->Read = false;
             $discussion->CountUnreadComments = true;
         } else {
+            // If the category was marked explicitly read at some point, see if that applies here
             if ($category && !is_null($category['DateMarkedRead'])) {
-                // If the category was marked explicitly read at some point, see if that applies here
-                $discussion->DateLastViewed = self::maxDate($discussion->DateLastViewed, $category['DateMarkedRead']);
+                // If the discussion hasn't been viewed or was created after the category was marked read,
+                // leave CountCountCommentWatch and DateLastViewed null. Otherwise, calculate the correct DateLastViewed.
+                if (!is_null($discussion->DateLastViewed) |
+                    self::maxDate($category['DateMarkedRead'], $discussion->DateInserted) === $category['DateMarkedRead']) {
```

Comment:
The `DiscussionModel::maxDate()` function specifies that it can take and return null values, but you do the check here. Wouldn't it be better to do the fix within `maxDate()` itself?

Reply:
I don't think I understand this comment, @tburry. `maxDate()` will only return null if both values passed to it are null. But that's not the case here. We don't want a null value to get overwritten with a date in this case. Can you explain what I'm missing?